src/bulkUtils.ts

```diff
@@ -174,3 +182,125 @@ export const remainingTime =
   (now: number) =>
   (endWaitTime?: number): number =>
     Math.max((endWaitTime ?? now) - now, 0);
+
+export enum ColumnDelimiter {
+  BACKQUOTE = '`',
+  CARET = '^',
+  COMMA = ',',
+  PIPE = '|',
+  SEMICOLON = ';',
+  TAB = '	',
+}
+
+export type ColumnDelimiterKeys = keyof typeof ColumnDelimiter;
+
+async function bulkRequest(conn: Connection, url: string): Promise<{ body: string; headers: HttpResponse['headers'] }> {
+  const httpApi = new HttpApi(conn, {
+    responseType: 'text/plain', // this ensures jsforce doesn't try parsing the body
+  });
+
+  let headers: HttpResponse['headers'] | undefined;
+
+  httpApi.on('response', (response: HttpResponse) => {
+    headers = response.headers;
+  });
+
+  const body = await httpApi.request<string>({
+    url: conn.normalizeUrl(url),
+    method: 'GET',
+  });
+
+  if (!headers) throw new Error('failed to get HTTP headers for bulk query');
+
+  return {
+    body,
+    headers,
+  };
+}
```

Comment:
small wrapper around jsforce's `HttpApi` class to be able to make requests and get response body **and** body.

`conn.request` only returns the body, we need the `sforce-locator` header to fetch the next batch.